client.go

```diff
@@ -120,9 +132,19 @@ func readExpr() <-chan expr {
 		s := bufio.NewScanner(c)
 		for s.Scan() {
 			log.Printf("recv: %s", s.Text())
-			p := newParser(strings.NewReader(s.Text()))
-			if p.parse() {
-				ch <- p.expr
+			if word, rest := splitWord(s.Text()); word == "recv" {
+				gState.mutex.Lock()
+				state, ok := gState.data[rest]
+				gState.mutex.Unlock()
+				if ok {
+					fmt.Fprint(c, state)
+				}
+				fmt.Fprintln(c, "")
+			} else {
+				p := newParser(strings.NewReader(s.Text()))
```

Comment:
Perhaps add a comment explaining why `recv` is not processed together with all the other commands?

My guess is that you're worried about some race condition.

Reply:
This is because processing a command the normal way will forward it to the main thread loop, which is blocked when running a `$`/`!` synchronous shell command. I have added a comment to explain this in f04e81b.